Adding more configuration comments.

diff --git a/debian/changelog b/debian/changelog
index 8280e36..6dc03ad 100644
--- a/debian/changelog
+++ b/debian/changelog
@@ -1,3 +1,9 @@
+isc-dhcp (4.2.2.dfsg.1-5+deb70u6+cl2) CumulusLinux-2.5; urgency=low
+
+  * Cumulus Networks patches
+
+ -- John Berezovik <johnberezovik@monster-01.cumulusnetworks.com>  Fri, 27 Mar 2015 16:04:23 -0700
+
 isc-dhcp (4.2.2.dfsg.1-5+deb70u6) stable-proposed-updates; urgency=medium
 
   * Set --with-ldapcrypto to restore openssl support (closes: #692808).
diff --git a/debian/isc-dhcp-relay.postinst b/debian/isc-dhcp-relay.postinst
index c75aa96..51960ac 100644
--- a/debian/isc-dhcp-relay.postinst
+++ b/debian/isc-dhcp-relay.postinst
@@ -44,6 +44,7 @@ INITCONFFILE=/etc/default/isc-dhcp-relay
 SERVERS="172.16.0.10"
 
 # On what interfaces should the DHCP relay (dhrelay) serve DHCP requests?
+# Always include the interface towards the DHCP server.
 INTERFACES="eth0"
 
 # Additional options that are passed to the DHCP relay daemon?
